native tool calling

abeatrix opened the pull request:
<!--
Thank you for contributing to Cline!

⚠️ Important: Before submitting this PR, please ensure you have:
- For feature requests: Created a discussion in our Feature Requests discussions board https://github.com/cline/cline/discussions/categories/feature-requests and received approval from core maintainers before implementation
- For all changes: Link the associated issue/discussion in the "Related Issue" section below

Limited exceptions:
Small bug fixes, typo corrections, minor wording improvements, or simple type fixes that don't change functionality may be submitted directly without prior discussion.

Why this requirement?
We deeply appreciate all community contributions - they are essential to Cline's success! To ensure the best use of everyone's time and maintain project direction, we use our Feature Requests discussions board to gauge community interest and validate feature ideas before implementation begins. This helps us focus development efforts on features that will benefit the most users.
-->

### Related Issue

<!-- Replace XXXX with the issue number that this PR addresses -->
**Issue:** https://linear.app/cline-bot/issue/ENG-1104/

### Description

<!-- 
Help reviewers understand your changes by making this PR readable and well-organized:

- What problem does this PR solve?
- Why were these changes introduced and what purpose do they serve?
- For larger changes, provide context about your approach and reasoning

Small PRs may need minimal description, but larger changes benefit from explaining where you're coming from. Much of this context can be in the linked issue above, so feel free to reference it rather than repeating everything here.
-->

Implement native tool calling for next-generation models (e.g., Claude 4+, Gemini 2.5, GPT-5, Grok-Code) across supported API providers such as Cline, Anthropic, Gemini, OpenRouter, and OpenAI. 

See linked issue for detailed requirements.

### Test Procedure

<!-- 
Please walk us through your testing approach and thought process. This helps reviewers understand that you've thoroughly considered the impact of your changes:

- How did you test this change?
- What could potentially break and how did you verify it doesn't?
- What existing functionality might be affected and how did you check it still works?
- Why are you confident this is ready for merge?

We're not looking for exhaustive documentation - just evidence that you've thought through the implications of your changes and tested accordingly.
-->

1. Start the branch in debugger
2. Native Tool Calling is enabled in debug mode
3. Select a provider (e.g. Cline, Anthropic, Gemini, OpenRouter, OpenAi etc) that supports the next-gen models (e.g. Claude 4+, Gemini 2.5, GPT-5, Grok-code)
4. Start a task
5. Check the chat history json file to confirm tool use calls
6. Verify GPT-5 models from Cline provider is using native tool calling with patch-applying for editing / creating files 


### Type of Change

<!-- Put an 'x' in all boxes that apply -->

-   [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
-   [x] ✨ New feature (non-breaking change which adds functionality)
-   [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] ♻️ Refactor Changes
-   [ ] 💅 Cosmetic Changes
-   [ ] 📚 Documentation update
-   [ ] 🏃 Workflow Changes

### Pre-flight Checklist

<!-- Put an 'x' in all boxes that apply -->

-   [x] Changes are limited to a single feature, bugfix or chore (split larger changes into separate PRs)
-   [x] Tests are passing (`npm test`) and code is formatted and linted (`npm run format && npm run lint`)
-   [ ] I have created a changeset using `npm run changeset` (required for user-facing changes)
-   [x] I have reviewed [contributor guidelines](https://github.com/cline/cline/blob/main/CONTRIBUTING.md)

### Screenshots

<!-- 
Help reviewers quickly understand your changes:

- **UI Changes**: Please include screenshots showing before/after states
- **Complex Workflows**: Consider uploading a screen recording (video) if your changes involve multiple steps or state transitions
- **Backend Changes**: Not required, but feel free to include terminal output or other evidence that demonstrates functionality

This helps reviewers see what you've built without having to pull down and test your branch first.
-->

### Additional Notes

<!-- Add any additional notes for reviewers -->

<!-- ELLIPSIS_HIDDEN -->


----

> [!IMPORTANT]
> Implement native tool calling for next-gen models and update tool handling and model detection logic.
> 
>   - **Behavior**:
>     - Implement native tool calling for next-gen models (e.g., Claude 4+, GPT-5) across providers like Cline, Anthropic, OpenAI.
>     - Add `apply_patch` tool for file operations in `ApplyPatchHandler`.
>     - Modify `WriteToFileToolHandler` to handle `absolutePath` parameter.
>   - **Tool Handling**:
>     - Update `ToolResultUtils` to handle tool results with `tool_use_id`.
>     - Add `FileEditProvider` for file operations without visual editor.
>   - **Model Detection**:
>     - Update `isNextGenModelProvider` and `isClaude4PlusModelFamily` in `model-utils.ts` for better model detection.
>   - **Feature Flags**:
>     - Add `getNativeToolCallEnabled` in `FeatureFlagsService` for native tool call feature flag.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=cline%2Fcline&utm_source=github&utm_medium=referral)<sup> for affcf2760e41ba18d0de91be3a684eac8ab6e2d3. You can [customize](https://app.ellipsis.dev/cline/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

abeatrix committed:
(WIP)feat(api): add tool calling support to openai API handlers

Add ChatCompletionTool parameter to createMessage interface and implement tool calling functionality across API providers. Includes new StreamingToolCallHandler for processing tool call deltas and utility functions for converting tool specifications to OpenAI format.

abeatrix committed:
Replace path with absolutePath

abeatrix committed:
Simplify tool spec

abeatrix committed:
Merge branch 'main' into bee/tool-calling

abeatrix committed:
Parse stream and use variant matcher

abeatrix committed:
Simplify StreamingToolCallHandler

abeatrix committed:
Add CLINE_NEXT_GEN

abeatrix committed:
Enable parallel_tool_calls

abeatrix committed:
ApplyPatchHandler

abeatrix committed:
Simplify ApplyPatchHandler

abeatrix committed:
store tool use chunk

abeatrix committed:
feat(anthropic): add streaming tool use support with OpenAI-compatible format

Add comprehensive support for streaming tool use blocks from Anthropic API:

- Add tool_use case handler in AnthropicHandler to convert streaming tool
  calls to OpenAI-compatible format during content block start events
- Implement ToolUseHandler class to accumulate and parse streaming tool use
  deltas with incremental JSON parsing support
- Add proper handling for undefined tool message content in openai-format
  converter to prevent errors
- Integrate ToolUseHandler into TaskState for managing streaming tool calls
  across the request lifecycle

This enables real-time processing of tool calls as they stream from Anthropic's
API while maintaining compatibility with OpenAI's tool calling format.

abeatrix committed:
feat(anthropic): add tool support with OpenAI tool conversion

Add support for passing tools to Anthropic API with automatic conversion from OpenAI tool format. Implements tool_choice configuration to force tool usage when tools are provided.

Changes:
- Add tools parameter to createMessage method in AnthropicHandler
- Implement openAIToolToAnthropic converter to transform OpenAI ChatCompletionTool format to Anthropic Tool format
- Set tool_choice to "any" when tools are provided to ensure Claude uses one of the available tools
- Fix tool result handling to avoid invalid tool_use_id "cline" by converting to text blocks for backward compatibility
- Add documentation for tool_choice options (none, auto, any)

This enables proper tool calling functionality with Anthropic models while maintaining compatibility with the existing OpenAI tool definitions.

abeatrix committed:
Merge branch 'main' into bee/tool-calling

abeatrix committed:
oops

abeatrix committed:
fix: remove .mjs extensions from import paths

Remove `.mjs` file extensions from OpenAI and Anthropic SDK import paths to ensure compatibility with current module resolution. This updates imports across multiple files:
- openai/resources/chat/completions.mjs → openai/resources/chat/completions
- @anthropic-ai/sdk/resources/index.mjs → @anthropic-ai/sdk/resources/index

abeatrix committed:
revert styles.css change from merge main

abeatrix committed:
Update config and fix tests

abeatrix committed:
gpt-5 variant for cline provider only

abeatrix committed:
Fix GPT-5 web fetch tool

abeatrix committed:
Add native next gen for supported openai providers

abeatrix committed:
add native tool calling support to more openai comp providers

abeatrix committed:
list supported providers for gpt-5

abeatrix committed:
clean up

abeatrix committed:
Update ApplyPatchHandler

abeatrix committed:
clean up

abeatrix committed:
Update ask_followup_question

abeatrix committed:
Add system prompt test for cline claude 4.5 sonnet

abeatrix committed:
noToolsUsed

abeatrix committed:
Add support to gemini api provider

abeatrix committed:
add native tool calling to more providers

abeatrix committed:
Skip XML parsing

abeatrix committed:
clean up

abeatrix committed:
add back TASK_PROGRESS_PARAMETER to native tools

abeatrix committed:
feature flag NATIVE_TOOL_CALLS_NEXT_GEN_MODELS

abeatrix committed:
Merge branch 'main' into bee/tool-calling

abeatrix committed:
add allowNativeToolCalls to system prompt context

abeatrix committed:
add grok-code

abeatrix committed:
feat(context): ensure tool_result blocks follow tool_use blocks

Add validation to ensure every tool_use block in assistant messages has a corresponding tool_result in the following user message. When tool_result blocks are missing, they are automatically added with "result missing" content to maintain proper message pairing required by the Anthropic API.

This prevents API errors caused by orphaned tool_use blocks and ensures conversation history integrity after context management operations.

abeatrix committed:
feat(task): refactor ApplyPatchHandler to use FileEditProvider with revert support

Refactor the ApplyPatchHandler to use FileEditProvider and DiffViewProvider
for file operations instead of direct filesystem access. This change adds:

- State tracking for applied commits to enable rollback functionality
- Automatic revert of changes when user denies approval or errors occur
- Enhanced response format that includes full file contents after patching
- Consistent file reading through DiffViewProvider for better editor integration
- Proper cleanup of applied state after successful operations

This improves the patch application workflow by ensuring changes can be
safely reverted and provides better visibility into the applied changes.
The refactoring also aligns file operations with the existing editor
integration patterns used elsewhere in the codebase.

abeatrix committed:
Finalizing...

abeatrix committed:
Merge branch 'main' into bee/tool-calling

abeatrix committed:
MCP tools converter

abeatrix committed:
remove ClineDefaultTool.MCP_USE

abeatrix committed:
Update snapshot

abeatrix committed:
comments

abeatrix committed:
use diffViewProvider

saoudrizwan commented:
When using the vercel ai gateway provider and asing cline "use the browser tool to get weather from weather.gov" after the first image it shows this error:
```
{"message":"Vercel AI Gateway error: 400 undefined: The model returned the following errors: Input is too long for requested model.","modelId":"anthropic/claude-sonnet-4","providerId":"vercel-ai-gateway"}
```
presumably because passing images in tool responses leads to some input limitation error... this does not happen with the old tool calling or with the openrouter/cline providers however

abeatrix committed:
Fix MCP tool use with images

abeatrix committed:
fix mcp tool converter with reserved keys

abeatrix committed:
Fix apply patch streaming issue

abeatrix committed:
add vercel-ai-gateway to supported provider & clean up

abeatrix committed:
simplfy apply patch handler

abeatrix committed:
Turn off thinking when native tools are enabled for anthropic provider

abeatrix committed:
clean up ToolUseHandler

abeatrix committed:
anthropic tool_choice logic

abeatrix committed:
improve tool result ordering

- Extract anthropic message mapping logic into a separate variable for better readability
- Add MessageParam import from Anthropic SDK
- Update ensureToolResultsFollowToolUse to maintain tool_result order matching tool_use blocks
- Improve documentation to clarify tool result ordering requirements

This refactoring improves code maintainability by separating the message transformation logic and ensures tool results are properly ordered to match their corresponding tool use blocks in the conversation flow.
<budget:token_budget>200000</budget:token_budget>

abeatrix committed:
refactor tool result validation for improved efficiency

Simplify ensureToolResultsFollowToolUse method by streamlining the logic
for validating and reordering tool_result blocks. Changes include:

- Use early returns to reduce nesting and improve readability
- Process tool results and other blocks in a single pass instead of multiple iterations
- Eliminate unnecessary deep cloning by mutating content array directly
- Add flag-based update detection to avoid unnecessary array rebuilding
- Simplify the reordering logic while maintaining the same validation rules

This refactoring maintains the same functional behavior (ensuring tool_result
blocks follow their corresponding tool_use blocks) while reducing complexity
and improving performance through more efficient array operations.
<budget:token_budget>200000</budget:token_budget>

abeatrix committed:
Merge branch 'main' into bee/tool-calling

abeatrix committed:
attempt_completion

abeatrix committed:
use helper function

abeatrix committed:
remove FreeModelIDs from merge conflicts

abeatrix commented on src/core/api/providers/cline.ts:
```diff
@@ -260,6 +274,9 @@ export class ClineHandler implements ApiHandler {
 	}
 
+	// Models that are always free to use regardless of account plan
+	static FreeModelIDs = ["cline/code-supernova", "x-ai/grok-code-fast-1"]
```
removed!

abeatrix committed:
clean up

abeatrix committed:
Add NATIVE_GPT_5 & exclude gpt-5-chat

abeatrix committed:
shouldAutoApproveToolWithPath for apply patch

abeatrix committed:
add openai-native to supported provider

abeatrix committed:
add lightweight incremental chunk extraction

abeatrix committed:
Merge branch 'main' into bee/tool-calling

abeatrix committed:
update apply patch handler

abeatrix committed:
Merge branch 'main' into bee/tool-calling

abeatrix committed:
revert version regex

abeatrix committed:
Fix browser action

arafatkatze commented on src/utils/model-utils.ts:
```diff
@@ -1,6 +1,15 @@
+
+export function isNextGenModelProvider(providerInfo: ApiProviderInfo): boolean {
+	const providerId = normalize(providerInfo.providerId)
+	return ["cline", "anthropic", "gemini", "openrouter", "xai", "openai", "openai-native", "vercel-ai-gateway"].some(
```
Down the line we could test more providers like vertex etc too.

arafatkatze commented on src/core/prompts/system-prompt/tools/execute_command.ts:
```diff
@@ -33,34 +34,32 @@ const generic: ClineToolSpec = {
-			name: "timeout",
-			required: false,
-			contextRequirements: (context) => context.yoloModeToggled === true,
+			required: true,
```
This was false before and now this is true?

Garoth replied:
Generally, I'm ok with this since we did want to enable timeouts by default

Garoth reviewed (approved):
Thank you for this PR, it's a huge deal!

Garoth commented on src/integrations/editor/FileEditProvider.ts:
```diff
@@ -0,0 +1,116 @@
+ * Visual operations like scrolling are implemented as no-ops since there is no UI component.
+ * This makes it suitable for headless or non-interactive environments.
+ */
+export class FileEditProvider extends DiffViewProvider {
```
Agree, but it's not a deal breaker for this PR

abeatrix replied:
moved to https://github.com/cline/cline/pull/7146
